pyinfra/api/connectors/winrm.py

```diff
@@ -0,0 +1,188 @@
+    hostname = kwargs.pop(
+        'hostname',
+        host.data.ssh_hostname or host.name,
+    )
+
+    logger.debug('winrm_username:{}'.format(host.data.winrm_username))
+    logger.debug('winrm_password:{}'.format(host.data.winrm_password))
+    logger.debug('winrm_port:{}'.format(host.data.winrm_port))
+    logger.debug('use_shell:{}'.format(host.data.use_shell))
+
+    try:
+        # Create new session
+        host_and_port = '{}:{}'.format(hostname, host.data.winrm_port)
+        logger.debug('host_and_port: {}'.format(host_and_port))
+
+        session = winrm.Session(
+            host_and_port,
+            auth=(
+                host.data.winrm_username,
+                host.data.winrm_password,
+            ),
+        )
+
+        return session
+
+    # TODO: add exceptions here
+    except Exception as e:
+        auth_kwargs = {}
+
+        for key, value in kwargs.items():
+            if key in ('username', 'password'):
+                auth_kwargs[key] = value
+                continue
+
+        auth_args = ', '.join(
+            '{0}={1}'.format(key, value)
+            for key, value in auth_kwargs.items()
+        )
+        logger.debug(str(e))
+        _raise_connect_error(host, 'Authentication error', auth_args)
+
+
+def run_shell_command(
+    state, host, command,
+    get_pty=False,
+    timeout=None,
+    stdin=None,
+    success_exit_codes=None,
+    print_output=False,
+    return_combined_output=False,
+    **command_kwargs
+):
+    '''
+    Execute a command on the specified host.
+
+    Args:
+        state (``pyinfra.api.State`` obj): state object for this command
+        hostname (string): hostname of the target
+        command (string): actual command to execute
+        get_pty (boolean): Not used for WINRM
+        env (dict): envrionment variables to set
+        timeout (int): timeout for this command to complete before erroring
+
+    Returns:
+        tuple: (exit_code, stdout, stderr)
+        stdout and stderr are both lists of strings from each buffer.
+    '''
+
+    #_shell_executable = 'cmd'
+    #if host.data.use_shell:
+        #_shell_executable = host.data.use_shell
+    command = make_command(command, **command_kwargs)
+
+    logger.debug('Running command on {0}: {1}'.format(host.name, command))
+
+    # get rid of leading/trailing quote
+    tmp_command = command.strip("'")
+
+    if print_output:
+        click.echo('{0}>>> {1}'.format(host.print_prefix, command))
+
+    if host.data.use_shell == 'cmd':
+        response = host.connection.run_cmd(tmp_command)
+    else:
+        response = host.connection.run_ps(tmp_command)
+
+    return_code = response.status_code
+
+    logger.debug('response:{}'.format(response))
+    logger.debug('response.status_code:{}'.format(response.status_code))
+    logger.debug('response.std_out:{}'.format(response.std_out))
+    logger.debug('response.std_err:{}'.format(response.std_err))
```

Comment:
These debug logs should be merged into one call; same with above too.

Reply:
I just temporarily left that in there while I was trying to figure things out. Removed.